ndb/src/google/cloud/ndb/query.py

```diff
@@ -134,28 +141,376 @@ def __init__(self, *args, **kwargs):
+    """
+
+    def __new__(cls, prop, op, param):
+        if not isinstance(prop, model.Property):
+            raise TypeError("Expected a Property, got {!r}".format(prop))
+        if op not in _OPS:
+            raise TypeError("Expected a valid operator, got {!r}".format(op))
+        if not isinstance(param, ParameterizedThing):
+            raise TypeError(
+                "Expected a ParameterizedThing, got {!r}".format(param)
+            )
+        obj = super(ParameterNode, cls).__new__(cls)
+        obj._prop = prop
+        obj._op = op
+        obj._param = param
+        return obj
+
+    def __getnewargs__(self):
+        """Private API used to specify ``__new__`` arguments when unpickling.
+
+        .. note::
+
+            This method only applies if the ``pickle`` protocol is 2 or
+            greater.
+
+        Returns:
+            Tuple[.Property, str, .ParameterizedThing]: A tuple containing the
+            internal state: the property, operation and parameter.
+        """
+        return self._prop, self._op, self._param
+
+    def __repr__(self):
+        return "ParameterNode({!r}, {!r}, {!r})".format(
+            self._prop, self._op, self._param
+        )
+
+    def __eq__(self, other):
+        if not isinstance(other, ParameterNode):
+            return NotImplemented
+        return (
+            self._prop._name == other._prop._name
+            and self._op == other._op
+            and self._param == other._param
+        )
+
+    def _to_filter(self, post=False):
+        """Helper to convert to low-level filter, or :data:`None`.
+
+        Args:
+            post (bool): Indicates if this is a post-filter node.
+
+        Raises:
+            .BadArgumentError: Always. This is because this node represents
+            a parameter, i.e. no value exists to be filtered on.
+        """
+        raise _exceptions.BadArgumentError(
+            "Parameter :{} is not bound.".format(self._param.key)
+        )
+
+    def resolve(self, bindings, used):
+        """Return a node with parameters replaced by the selected values.
+
+        Args:
+            bindings (dict): A mapping of parameter bindings.
+            used (Dict[Union[str, int], bool]): A mapping of already used
+                parameters.
+
+        Raises:
+            NotImplementedError: Always. This is because the implementation
+                will rely on as-yet-unimplemented features in
+                :class:`Property`.
+        """
+        raise NotImplementedError(
+            "Some features of Property need to be implemented first"
+        )
 
 
 class FilterNode(Node):
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    """Tree node for a single filter expression.
+
+    For example ``FilterNode("a", ">", 3)`` filters for entities where the
+    value ``a`` is greater than ``3``.
+
+    .. warning::
+
+        The constructor for this type may not always return a
```

Comment:
Note this warning. These "constructors" are quite devious. (I made sure to mention this in the `MIGRATION_NOTES.md`)

Reply:
Luckily, these `Node` classes are never really instantiated in user code. However, this behavior has to be preserved (at least for now) for the migration. A node is typically created via comparison to a property:

```python
filter_ = SomeModel.total > 1000
```

Reply:
Agreed. I'd prefer to take the conservative approach here, if everyone else is on board. The goal was "full interface compatibility".